### PR TITLE
Signing tools

### PR DESCRIPTION
**Pull request by oskin1**

### CODE REVIEW

**Comment by oskin1** [comment]
Third task from #1

**Comment by oskin1** [comment]
Relies on #3

File: src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    val storageFeeCovered = box.value - maxStorageFee <= 0
+    val correctCreationHeight = output.creationHeight == currentHeight
+    val correctOutValue = output.value >= box.value - maxStorageFee
+    val correctRegisters = ErgoBox.allRegisters.tail
+      .forall(rId => rId == ErgoBox.ReferenceRegId || box.get(rId) == output.get(rId))
+
+    storageFeeCovered || (correctCreationHeight && correctOutValue && correctRegisters)
+  }
+
+  override def verify(env: ScriptEnv,
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala]: Please add comments describing parameters

File: src/main/scala/org/ergoplatform/wallet/interpreter/ErgoProvingInterpreter.scala [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                            (implicit IR: IRContext)
+  extends ErgoInterpreter(params) with ProverInterpreter {
+
+  val pubKeys: IndexedSeq[ProveDlog] = secrets.map(_.publicImage)
+
+  /** Require `unsignedTx` and `boxesToSpend` have the same boxIds in the same order */
+  def sign(unsignedTx: UnsignedErgoLikeTransaction,
+           boxesToSpend: IndexedSeq[ErgoBox],
+           dataBoxes: IndexedSeq[ErgoBox],
+           stateContext: ErgoLikeStateContext): Try[ErgoLikeTransactionTemplate[Input]] = {
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/wallet/interpreter/ErgoProvingInterpreter.scala]: The result should be Try[ErgoLikeTransaction]

File: src/main/scala/org/ergoplatform/wallet/protocol/context/ErgoContext.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package org.ergoplatform.wallet.protocol.context
+
+import org.ergoplatform.wallet.protocol.Constants
+import org.ergoplatform.{ErgoLikeContext, ErgoLikeTransactionTemplate, UnsignedInput}
+import scorex.crypto.authds.ADDigest
+import sigmastate.interpreter.ContextExtension
+import sigmastate.{AvlTreeData, AvlTreeFlags}
+
+class ErgoContext(val stateContext: ErgoLikeStateContext,
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/wallet/protocol/context/ErgoContext.scala]: Please add class comments

Comment (catena2w) [inline src/main/scala/org/ergoplatform/wallet/protocol/context/ErgoContext.scala]: Is it possible to simplify this class to reduce the number of classes with context? Do we really need to distinct `ErgoLikeStateContext` and `TransactionContext`? Or maybe we don't even need `ErgoContext` class here and can use ErgoLikeContext from sigma?

**Review by catena2w — CHANGES_REQUESTED** [review]

**Review by catena2w — APPROVED** [review]

**Pull request merged**